---
title: "Preprocess data"
output: html_notebook
---

## Setup

```{r}
library(ggplot2)
library(magrittr)

source("../R/preprocessing.R")

training_set <- read.csv("../training_data.csv", as.is = TRUE)
training_set$outcome <- compute_numeric_labels(training_set$CLNSIG)
```

## Choose features

The features are selected to be values from tools in dbNSFP that are not themselves already metapredictors. E.g. MetaSVM and Eigen are thus filtered out. From CADD annotations, features are chosen by using some intuition of whether they might be usable by the classifier.

```{r}
numeric_features <- make.names(c(
  # From dbNSFP
  "SIFT_score", 
  "LRT_score", 
  "LRT_Omega", 
  "MutationTaster_score", 
  "MutationAssessor_score", 
  "FATHMM_score", 
  "PROVEAN_score", 
  "M-CAP_score", 
  "MutPred_score", 
  "fathmm-MKL_coding_score", 
  "GenoCanyon_score", 
  "integrated_fitCons_score", 
  "GERP++_NR", 
  "GERP++_RS", 
  "phyloP100way_vertebrate", 
  "phyloP20way_mammalian", 
  "phastCons100way_vertebrate", 
  "phastCons20way_mammalian", 
  "SiPhy_29way_logOdds", 
  "gnomAD_exomes_AF",
  "gnomAD_genomes_AF",
  
  # From CADD annotations
  "Length",
  "GC",
  "CpG",
  "motifECount",
  "motifEHIPos",
  "motifEScoreChng",
  # "(Intron)",
  # "(Exon)",
  "relcDNApos",
  "relCDSpos",
  "relProtPos",
  "Dst2Splice",
  "minDistTSS",
  "minDistTSE",
  # "targetScan",
  "mirSVR-Score",
  "mirSVR-E",
  "mirSVR-Aln",
  "tOverlapMotifs",
  "motifDist",
  "EncodeH3K4me1-sum",
  "EncodeH3K4me1-max",
  "EncodeH3K4me2-sum",
  "EncodeH3K4me2-max",
  "EncodeH3K4me3-sum",
  "EncodeH3K4me3-max",
  "EncodeH3K9ac-sum",
  "EncodeH3K9ac-max",
  "EncodeH3K9me3-sum",
  "EncodeH3K9me3-max",
  "EncodeH3K27ac-sum",
  "EncodeH3K27ac-max",
  "EncodeH3K27me3-sum",
  "EncodeH3K27me3-max",
  "EncodeH3K36me3-sum",
  "EncodeH3K36me3-max",
  "EncodeH3K79me2-sum",
  "EncodeH3K79me2-max",
  "EncodeH4K20me1-sum",
  "EncodeH4K20me1-max",
  "EncodeH2AFZ-sum",
  "EncodeH2AFZ-max",
  "EncodeDNase-sum",
  "EncodeDNase-max",
  "EncodetotalRNA-sum",
  "EncodetotalRNA-max",
  # "Grantham",
  "RemapOverlapTF",
  "RemapOverlapCL"
))

categorical_features <- make.names(c(
  # From dbNSFP
  "LRT_pred",
  
  # From CADD annotations
  "Type",
  "Dst2SplType",
  "Consequence.x"
))

training_data <- training_set[, c(numeric_features, categorical_features, "outcome")]
```

## Descriptive statistics

- Check whether all consequences have both positive and negative examples
  - they don't.
```{r}
table_with_margin <- function(...) {
  tabl <- table(...)
  tabl %<>% cbind(ALL_ = rowSums(tabl))
  tabl %<>% rbind(ALL_ = colSums(tabl))
  return(tabl)
}
conseq_by_clnsig <- table_with_margin(training_set$Consequence.x, training_set$CLNSIG, useNA = "always")
conseq_by_outcome <- table_with_margin(training_set$Consequence.x, training_set$outcome, useNA = "always")
conseq_by_clnsig
conseq_by_outcome
```
- Plot distributions of each feature. Are they normal or linear?
```{r}
invisible(lapply(numeric_features, function(column) plot(ggplot2::quickplot(na.omit(training_data[,column]), main = column, bins = 30))))
```


```{r}
lapply(training_data[, categorical_features], function(x) table(x, useNA = "always"))
```
```{r}
nas <- lapply(training_data, function(x) sapply(split(is.na(x), training_data$Consequence), sum)) 
nas %<>% data.frame %>% as.matrix
heatmap(nas)
```
```{r}
num_missingness_patterns <- training_data %>% is.na %>% unique %>% nrow
print(paste(num_missingness_patterns, "out of", 2^ncol(training_data), "possible missingness patterns."))
```

```{r}
training_data <- cbind(
  training_data[, numeric_features, drop = FALSE], 
  dummify_categoricals(training_data[, categorical_features, drop = FALSE]),
  outcome = training_data$outcome
)

write.csv(training_data, "../preprocessed_training_data.csv", row.names = FALSE)
```

